ree/reeEstimateError.cc

```diff
@@ -0,0 +1,458 @@
+  int nodesOnEdges = fs->countNodesOn(apf::Mesh::EDGE) * nedges;
+
+  apf::Downward faces;
+  int nfaces = mesh->getDownward(e, 2, faces);
+  PCU_ALWAYS_ASSERT(nfaces == 4);
+  for (int i = 0; i < nfaces; i++) {
+    int nodesOnFace = fs->countNodesOn(mesh->getType(faces[i]));
+    if ( crv::isBoundaryEntity(mesh, faces[i]) ) {
+      for (int n = 0; n < nodesOnFace; n++) {
+        marker_list[(nodesOnEdges + i*nodesOnFace)+n] = -1;
+      }
+    }
+  }
+
+  int num_marked = 0;
+  for (int i = 0; i < nedofs; i++) {
+    if (marker_list[i])
+      num_marked++;
+  }
+
+  // use marker list to get ess_dofs list
+  ess_dofs.allocated() ? ess_dofs.resize(num_marked)
+    : ess_dofs.allocate(num_marked);
+  uness_dofs.allocated() ? uness_dofs.resize(nedofs-num_marked)
+    : uness_dofs.allocate(nedofs-num_marked);
+  int ess_dof_counter = 0;
+  int uness_dof_counter = 0;
+  for (int i = 0; i < nedofs; i++) {
+    if(marker_list[i])
+      ess_dofs[ess_dof_counter++] = i;
+    else
+      uness_dofs[uness_dof_counter++] = i;
+  }
+}
+
+/**
+ * Inputs: Matrix A, Vector X, Vector B, essential dofs, other dofs.
+ * Output: reduced matrix A, reduced rhs B.
+ */
+static void eliminateDBCs(
+    mth::Matrix<double> const &A,
+    mth::Vector<double> const &X,
+    mth::Vector<double> const &B,
+    apf::NewArray<int>  const &ess_dofs,
+    apf::NewArray<int>  const &uness_dofs,
+    mth::Matrix<double> &Anew,
+    mth::Vector<double> &Bnew)
+{
+  int num_ess_dofs = ess_dofs.size();
+  int num_uness_dofs = uness_dofs.size();
+
+  // 1. Remove rows of A corresponding to
+	// ess_dofs by copying into Anew
+  Anew.resize(num_uness_dofs, num_uness_dofs);
+	for(int rr = 0; rr < num_uness_dofs; rr++) {
+    int i = uness_dofs[rr];
+    for(int cc = 0; cc < num_uness_dofs; cc++) {
+      int j = uness_dofs[cc];
+      Anew(rr,cc) = A(i,j);
+    }
+  }
+
+	// 2. Assemble new B
+	Bnew.resize(num_uness_dofs);
+  for(int i = 0; i < num_uness_dofs; i++) {
+    Bnew(i) = B(uness_dofs[i]);
+  }
+
+  if (num_ess_dofs > 0) {
+	  // 3. Subtract from Bnew: (Bnew -= Ae*Xe)
+    mth::Matrix<double> Ae(num_uness_dofs, num_ess_dofs);
+    for(int rr = 0; rr < num_uness_dofs; rr++) {
+      int i = uness_dofs[rr];
+      for(int cc = 0; cc < num_ess_dofs; cc++) {
+        int j = ess_dofs[cc];
+        Ae(rr,cc) = A(i,j);
+      }
+    }
+
+    mth::Vector<double> Xe(num_ess_dofs);
+    for(int i = 0; i < num_ess_dofs; i++) {
+      Xe(i) = X(ess_dofs[i]);
+    }
+
+    mth::Vector<double> temp;
+    mth::multiply(Ae, Xe, temp);
+    Bnew -= temp;
+  }
+}
+
+static double computeL2Error(apf::Mesh* mesh, apf::MeshEntity* e,
+	apf::Field* f, mth::Vector<double> const error_dofs)
+{
+	double error = 0.0;
+
+  apf::FieldShape* fs = f->getShape();
+  int type = mesh->getType(e);
+	PCU_ALWAYS_ASSERT(type == apf::Mesh::TET);
+  int nd = apf::countElementNodes(fs, type);
+  int dim = apf::getDimension(mesh, e);
+  double w;
+
+  apf::MeshElement* me = apf::createMeshElement(mesh, e);
```

Comment:
@Samiullah-Malik 

each time you create a `MeshElement` (or `Element`) you have to call the `apf::destroyMeshElement(me)` (or `destroyElement(el)`) on that object before exiting the code block they are defined in. Otherwise you are going to have many memory leaks. For example for the ones defined here you would call the destroy apis after the for loop finishes on line 347.

Reply:
Resolved.